src/input/espresso.rs

```diff
@@ -1 +1,970 @@
+    node_index: HashMap<PubKey, usize>,
+
+    /// The delegators to each active node.
+    #[allow(dead_code)]
+    delegators: BTreeMap<Address, Vec<(Address, U256)>>,
+}
+
+impl EpochState {
+    /// Construct the epoch state from a stake table and DRB result.
+    fn new(number: u64, nodes: ValidatorMap, drb_result: DrbResult) -> Self {
+        // Get active node addresses.
+        let active_nodes = nodes.keys().copied().collect::<Vec<_>>();
+        // Index active nodes by staking key.
+        let node_index = nodes
+            .values()
+            .enumerate()
+            .map(|(i, node)| (node.stake_table_key, i))
+            .collect();
+
+        // Pre-process a randomized committee for leader election.
+        let entries = nodes
+            .values()
+            .map(|node| StakeTableEntry {
+                stake_key: node.stake_table_key,
+                stake_amount: node.stake,
+            })
+            .collect();
+        let committee = generate_stake_cdf(entries, drb_result);
+
+        // Index delegator lists by address.
+        let delegators = nodes
+            .into_values()
+            .map(|node| (node.account, node.delegators.into_iter().collect()))
+            .collect();
+
+        Self {
+            number,
+            committee,
+            active_nodes,
+            node_index,
+            delegators,
+        }
+    }
+
+    /// Fetch the state for an epoch from an Espresso query service.
+    #[instrument(skip(espresso))]
+    async fn download(
+        espresso: &impl EspressoClient,
+        epoch_height: u64,
+        epoch: u64,
+    ) -> Result<Self> {
+        // We require the first epoch at least to be completed, so we have a "previous" epoch from
+        // which we can get the stake table and DRB result for this epoch.
+        ensure!(
+            epoch > 1,
+            Error::internal().context("Espresso state must be started after epoch 1")
+        );
+
+        tracing::info!("fetching epoch state");
+        let stake_table = espresso
+            .stake_table_for_epoch(epoch)
+            .await
+            .map_err(|err| err.context("fetching stake table"))?;
+
+        // Get the DRB result from the transition block of the previous epoch.
+        let transition_block = transition_block_for_epoch(epoch - 1, epoch_height);
+        let drb_leaf = espresso
+            .leaf(transition_block)
+            .await
+            .map_err(|err| err.context(format!("fetching transition leaf {transition_block}")))?;
+        // Sanity check that we actually got a block from the previous epoch.
+        ensure!(
+            drb_leaf.epoch(epoch_height).as_deref().copied() == Some(epoch - 1),
+            Error::internal().context(format!(
+                "transition leaf {drb_leaf:?} is not from expected epoch {} (epoch_height={epoch_height})",
+                epoch - 1
+            ))
+        );
+        let drb_result = drb_leaf.next_drb_result.ok_or_else(|| {
```

Comment:
@ss-es  There was one leaf where this `next_drb_result` was not present. AFAIK it was only that one very old leaf and shouldn't create any issue for us but just want to be sure, because we recalculate drb in hotshot if `next_drb_result` is not present.

Reply:
If so I think it should be ok because we will be starting from the latest epoch and not replaying old data for this part

Reply:
yeah, so it shouldn't be a problem for us if its only that one leaf, but I don't know much about this drb thing so Salman can confirm